### PR TITLE
Avoid AOT CSE across backward dependency boundary

### LINKED ISSUE #161326: Different #kernels generated for fwd call with requires_grad=True/False
For the forward call of a function, running with `requires_grad=False` generates 3 kernels. But running with `requires_grad=True` generates 6 kernels. I assume this is due to saved tensor or graph partition. But saved tensors should be an extra output in the fwd call which does not need recomputation.

```
import torch
torch._inductor.config.force_disable_caches = True
# torch._inductor.config.triton.cooperative_reductions = True
@torch.compile(fullgraph=True, dynamic=False)
def _local_metrics(x):
    return x.abs().max(), x.abs().mean(), x.square().mean()
_local_metrics(torch.rand([512, 512, 512], device="cuda", requires_grad=True)) # <--- try switch between requires_grad=True/False
```

Slightly related issue: #161138

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184044

Keep AOTAutograd joint graph CSE from merging forward-only values with nodes needed by backward before min-cut partitioning. This preserves Inductor's reduction fusion for grad-enabled forwards while retaining CSE within each dependency class.

Fixes #161326
Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by bobrenjc93 — APPROVED** [review]
Consider verifying the backward pass produces correct gradients:

```
test/inductor/test_perf.py:395 — The new
  test_aot_autograd_cse_preserves_reduction_fusion test only
  exercises the forward pass (no .backward() call). This is
  sufficient for asserting forward kernel count, but doesn't
  verify that the backward pass still produces correct gradients
  after the CSE change. Adding result[0].backward() (and checking
  x.grad) or a torch.autograd.gradcheck would strengthen this as a
   regression test for #161326.
```

**Comment by jansel** [comment]
Agent, please verify that

**Comment by jansel** [comment]
My agent says I added compiled-vs-eager gradient verification to the regression test and validated the focused test locally.

**Comment by jansel** [comment]
@claude are those CI failures related to this PR?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Comment by jansel** [comment]
My agent says I updated the stale test_stream_backward_simple inline expectations for the intended duplicated mul after the AOT CSE dependency-boundary change, rebased the PR, and validated the focused stream test, neighboring stream sync test, original Inductor regression, multiprocess Transformers skip guard, and lintrunner.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?
